### PR TITLE
Add protocols and docs for registration

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Document the element registration metavars, in the spirit of #3076.

The fixup commit is to help reviewers.

We can also consider retrieving the list of registrars directly from the `Parallel::Phase::Registration` phase of the parallel component.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/Protocols/ElementRegistrar.hpp
Comment (knelli2) [inline src/Parallel/Protocols/ElementRegistrar.hpp]: Make these `\code ... \endcode` or actual `\snippet`s? Makes the docs easier to read.

**Review by knelli2 — CHANGES_REQUESTED** [review]
Just a small thing. Can be squashed in

**Review by nilsdeppe — COMMENTED** [review]
I have no additional comments. No need to wait for my explicit approval to merge.

File: src/Parallel/Protocols/ElementRegistrar.hpp
Comment (nilsvu) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: There's a `\snippet` below. I think this list of requirements is fine as explicit text.

**Comment by nilsvu** [comment]
Rebased

File: src/Parallel/Protocols/ElementRegistrar.hpp
Comment (knelli2) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: If you have the `\snippet` below, why do you need to rewrite the signature in a one-line comment? Just say "This protocol requires two functions, `perform_registration` and `perform_deregistration`, that follow the signatures given below in the example" or something to that effect.

In my opinion, the unformatted, single-color, single-line signature is hard to read.

Comment (nilsvu) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: Tried it, but then you have to look into the file where the snippet is to see the signature.

Comment (nilsvu) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: Switched to a code block now instead of the inline code

Comment (knelli2) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: > Tried it, but then you have to look into the file where the snippet is to see the signature.

Really? Then what does the code snippet actually show in the docs? Because I would have thought it showed the whole class (based on your changes) which should definitely have the signatures.

Comment (nilsvu) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: In the docs it prints the snippet, but in the code there's just the `\snippet` so you have to find it. But the code block prints well.

Comment (knelli2) [reply src/Parallel/Protocols/ElementRegistrar.hpp]: Oh ok, I see now. Guess this is dependent on how we want people to discover/use documentation. For devs, I feel it's usually easier to go to the source file and read there, so in that case explicitly writing out the signature is helpful. However, anybody who uses the website (maybe contributors) will now see duplicated signatures (or possibly out of sync signatures if we forget to update the docs). Unclear which we should make the norm right now but that discussion is not for this PR. This looks good for now.

**Review by knelli2 — APPROVED** [review]

**Pull request merged**
